### PR TITLE
dns_filter: add support for wildcard name in DNS filter inline table

### LINKED ISSUE #32941: Wildcard records support in DNS filter inline table
**Add support for wildcard records in DNS filter inline table**

I've tested locally and DNS filter doesn't seem to support wildcard records for inline table, e.g. snippet below does not match `abc.example.com` and forward request to configured resolvers.
```yaml
inline_dns_table:
  virtual_domains:
    - name: "*.example.com"
      endpoint:
        address_list:
          address:
          - 127.0.5.5
```

Didn't find any other available options to configure this in docs so I assume it is not supported yet.

### PR DESCRIPTION
**Pull request by StupidScience**
Commit Message: Add support for wildcard name in DNS filter inline table
Additional Description: 
Risk Level: Medium
Testing: I ran only unit tests relevant to dns_filter
Docs Changes: -
Release Notes: Added support wildcard names in :ref:`inline_dns_table
    <envoy_v3_api_field_extensions.filters.udp.dns_filter.v3.DnsFilterConfig.ServerContextConfig.inline_dns_table>`.
Platform Specific Features: N/A
Fixes #32941

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning Yanjun for a first-pass review as a code-owner:
/assign @yanjunxiang-google

**Comment by adisuissa** [comment]
Assigning to Yan for a review (after Yanjun's pass).
/assign @yanavlasov

**Comment by yanjunxiang-google** [comment]
Thanks @StupidScience for working on this!
Could you please fix the CI error?

**Comment by StupidScience** [comment]
@yanjunxiang-google sorry, original error that is exposed (on screenshot) did not give me a good idea on what's wrong (it was speling) so I thought it was some CI infra failure.
image

Should be fixed now.

**Comment by StupidScience** [comment]
@yanjunxiang-google all checks have passed finally

**Comment by alyssawilk** [comment]
@yanjunxiang-google PTAL

File: changelogs/current.yaml
Comment (yanjunxiang-google) [inline changelogs/current.yaml:231]: Can we change to this: Added support for wildcard resolving when DNS filter is working in server mode?

File: source/extensions/filters/udp/dns_filter/dns_filter.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/udp/dns_filter/dns_filter.cc:43]: Should we call this virtual domain name, or wildcard domain name?

File: source/extensions/filters/udp/dns_filter/dns_filter_utils.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/udp/dns_filter/dns_filter_utils.cc:92]: Should we call this function getWildcardDomainName?

File: source/extensions/filters/udp/dns_filter/dns_filter_utils.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/udp/dns_filter/dns_filter_utils.cc:99]: What if the domain_name is invalid, like just "*.".  This will return an empty string?

File: source/extensions/filters/udp/dns_filter/dns_filter.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/udp/dns_filter/dns_filter.cc:44]: Should we getDomainSuffix(virtual_domain.name()),
 or  getDomainSuffix( Utils::getVirtualDomainName(virtual_domain.name()))

Can these two return different result in some cases?

File: test/extensions/filters/udp/dns_filter/dns_filter_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/udp/dns_filter/dns_filter_test.cc:2439]: Can we add some tests for invalid virtual_domain test, like "*.com", "*.", "com", and see whether there is any issue?

File: source/extensions/filters/udp/dns_filter/dns_filter_utils.cc
Comment (StupidScience) [reply source/extensions/filters/udp/dns_filter/dns_filter_utils.cc:99]: If domain is `*.` it will return just `.`. You can see that in tests.
While working on these code I made few assumptions and cut the corners sometimes. Mostly because current code does not have any validation for passed domain names and in proto validations are: length >= 1 + header value regex, that's probably incorrect.
I did not want to increase the scope of my changes so decided not to cover validations. It is probably should be addressed, likely in different PR tho

File: source/extensions/filters/udp/dns_filter/dns_filter.cc
Comment (StupidScience) [reply source/extensions/filters/udp/dns_filter/dns_filter.cc:44]: Yeah, I thought about this (getting suffix of "initial" input domain name). But with current implementation result should always be the same.

**Comment by yanavlasov** [comment]
@StupidScience can you update the doc for the `virtual_domain ` api/envoy/data/dns/v3/dns_table.proto please? It needs to clarify semantics for how how wildcards are matched.

/wait

**Comment by StupidScience** [comment]
@yanavlasov done

**Comment by yanavlasov** [comment]
> @yanavlasov done

Thanks. The important bit is how many subdomains can be matched by the wildcard. Will *.foo.com match only bar.foo.com or also baz.bar.foo.com ? Can you clarify it in the comment, please?

**Comment by StupidScience** [comment]
@yanavlasov thanks for feedback. Added this info as well as info about matching precedence. Can you think of any other info that should be covered?

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
LGTM from me. Will wait for LGTM from @yanjunxiang-google and then submit

@StupidScience please resolve the merge error.

/wait

File: source/extensions/filters/udp/dns_filter/dns_filter_utils.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/udp/dns_filter/dns_filter_utils.cc:92]: It's okay to me then.

**Comment by yanjunxiang-google** [comment]
@StupidScience Could you please also add a couple of integration test cases here: https://github.com/envoyproxy/envoy/blob/f3ee7c778bce18db1862774d70cd36e019344847/test/extensions/filters/udp/dns_filter/dns_filter_integration_test.cc#L263

It will help show how the wildcard domain name works end-to-end.

**Comment by StupidScience** [comment]
@yanjunxiang-google added integration test, please check if that's smth you had in mind?

**Review by wbpcode — COMMENTED** [review]
/lgtm api

**Comment by yanjunxiang-google** [comment]
LGTM

**Review by yanjunxiang-google — APPROVED** [review]

**Comment by yanjunxiang-google** [comment]
> @yanjunxiang-google added integration test, please check if that's smth you had in mind?

Looks good. Thanks for adding this feature support!

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
LGTM. @StupidScience please resolve merge error.

/wait

**Comment by StupidScience** [comment]
@yanavlasov done

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
